lib/enumerables.rb

```diff
@@ -0,0 +1,164 @@
+    while i < arr.length
+      yield arr[i]
+      i += 1
+    end
+  end
+
+  # my_each_with_index
+  def my_each_with_index
+    return to_enum(:my_each_with_index) unless block_given?
+
+    arr = self
+    i = 0
+    while i < arr.length
+      yield arr[i], i
+      i += 1
+    end
+  end
+
+  # my_select
+  def my_select
+    return to_enum(:my_select) unless block_given?
+
+    temp_arr = []
+    arr = self
+    arr.my_each do |num|
+      ans = yield num
+      temp_arr.push(num) if ans == true
+    end
+    temp_arr
+  end
+
+  # my_all
+  def my_all?(paramet = nil)
+    arr = self
+    return true if arr.empty?
+
+    if paramet.nil? && block_given?
+      arr.my_each do |n|
+        ans = yield n
+        return false if ans == false
+      end
+    elsif paramet.nil? && !block_given?
+      arr.my_each { |n| return false if n.nil? || !n }
+    elsif paramet.is_a?(Regexp)
+      arr.my_each { |x| return false unless x.match(paramet) }
+    elsif paramet.is_a?(Module)
+      my_each { |x| return false if x.is_a?(paramet) }
+    else
+      my_each { |x| return false if x != paramet }
+    end
+    true
+  end
+
+  # my_any
+  def my_any?(paramet = nil)
+    arr = self
+    return true if arr.empty?
+
+    if paramet.nil? && block_given?
+      arr.my_each do |n|
+        ans = yield n
+        return true if ans == true
+      end
+    elsif paramet.nil? && !block_given?
+      arr.my_each { |n| return true if !!n == true && !n.nil? }
+    elsif paramet.is_a?(Regexp)
+      arr.my_each { |x| return true if x.match(paramet) }
+    elsif paramet.is_a?(Module)
+      my_each { |x| return true if x.is_a?(paramet) }
+    else
+      my_each { |x| return true if x == paramet }
+    end
+    false
+  end
+
+  # my_none
+  def my_none?(paramet = nil)
+    arr = self
+    return true if arr.empty?
+
+    if paramet.nil? && block_given?
+      arr.my_each do |n|
+        ans = yield n
+        return false if ans == true
+      end
+    elsif paramet.nil? && !block_given?
+      arr.my_each { |n| return false if !!n == true && !n.nil? }
+    elsif paramet.is_a?(Regexp)
+      arr.my_each { |x| return false if x.match(paramet) }
+    elsif paramet.is_a?(Module)
+      my_each { |x| return false if x.is_a?(paramet) }
+    end
+    true
+  end
+
+  # my_count
+  def my_count(para = nil)
+    arr = self
+    if para.nil? && !block_given?
+      counter = 0
+      arr.my_count { counter += 1 }
+      counter
+    elsif !para.nil? && !block_given?
+      counter = 0
+      arr.my_each { |x| counter += 1 if x == para }
+      counter
+    elsif para.nil? && block_given?
+      counter = 0
+      arr.my_each { |x| counter += 1 if yield x }
+      counter
+    end
+  end
+
+  # my_map
+  def my_map(proc = nil)
+    return to_enum unless block_given? || proc
+
+    arr = self
+    new_arr = []
+    if proc
+      arr.my_each { |num| new_arr.push(proc.call(num)) }
+    else
+      arr.my_each { |num| new_arr.push(yield(num)) }
+    end
+    new_arr
+  end
```

Comment:
These are the problems with `#my_map`

```
  17) enumerables #my_map returns a new array with the results of running a given block once for every element in enum::range
      Failure/Error: while i < arr.length
      
      NoMethodError:
        undefined method `length' for 5..50:Range
      # ./lib/enumerables.rb:8:in `my_each'
      # ./lib/enumerables.rb:130:in `my_map'
      # ./spec/enumerables_spec.rb:376:in `block (3 levels) in <top (required)>'
```

Again, please prepare this method to process ranges, not only arrays.